### DISCUSSION TITLE
How does CopyOperation async/sync logic work

### DISCUSSION THREAD

**Discussion by MattSom** [O365/python-o365#865] [Q&A]
In `drive.py` there is this method:

```python
def copy(self, target=None, name=None):
    """ Asynchronously creates a copy of this DriveItem and all it's
    child elements.
    
    ...
    
    response = self.con.post(url, data=data)
    if not response:
        return None

    # Find out if the server has run a Sync or Async operation
    location = response.headers.get('Location', None)

    if response.status_code == 202:
        # Async operation
        return CopyOperation(parent=self.drive, monitor_url=location)
    else:
        # Sync operation. Item is ready to be retrieved
        path = urlparse(location).path
        item_id = path.split('/')[-1]
        return CopyOperation(parent=self.drive, item_id=item_id)
```

My goal is to simply get back the `object_id` of the copied file. For this I need to call the `get_item()` on the returned `CopyOperation` object. Checking its code I see that it only works if that object has the `item_id` which is set only in with sync operation. This will call the parent drive's `get_item` with that id.

Otherwise there's some other `check_status` method which seems more of an async loop. Also the class has this comment as well:
https://github.com/OneDrive/onedrive-api-docs/issues/762

Does this mean that I can always expect the `item_id` to be existing and the `async` branch will never run? Do I need to prepare my code for this? Or can I just expect that the `monitor_url` will have the id:

```python
operation = drive_item_to_copy.copy(target=target_folder, name=target_path.name)

if copied_item_id := operation.get_item():
    return copied_item_id

path = urlparse(operation.monitor_url).path
return path.split("/")[-1]
 ```

**Comment by alejcas** [comment]
I don't know for sure. Can you check the MS Graph docs to see if it mentions something usefull?

**Reply by MattSom** [reply]
I will sure. Until that can you give some insight into the motivation of this code? I need that for the context to actually know what issue am I looking for and why was this written this way.

**Reply by alejcas** [reply]
Yes, I don't know now, but at the time I was writing that code, The MS Graph docs (which btw are sometimes misleading) mention the copy operation and all those endpoints to get the operation working.

I think that got changed some time after I wrote this. Also I can't tell exactly how, but I think even the MS GRAPH api changed it's behavior.
